text/0162-refactoring-support.md

```diff
@@ -0,0 +1,606 @@
+perceived as a workaround. However, we are open to implementing this as an
+additional feature if enough customers indicate their preference for it.
+
+### What are the drawbacks of this solution?
+
+See the open issues section below.
+
+### What is the high-level project plan?
+
+#### Phase 1 (dry-run)
+
+In this phase we are going to implement the detection of resource moves, and
+show the user what changes are going to be made. The only new command available
+at this phase is `cdk refactor --dry-run`. Execution of this command without the
+`--dry-run` flag will result in an error.
+
+High-level tasks:
+
+- Implement resource equivalence without physical ID.
+- Implement the computation of mappings.
+- Implement the display of the mappings to the user.
+- Implement ambiguity detection.
+- Implement the display of the ambiguities to the user.
+- Add physical ID to resource equivalence.
+
+#### Phase 2 (application)
+
+Once the detection of all cases is implemented in phase 1, we are ready to
+implement the application of the changes.
+
+High-level tasks:
+
+- Add new permissions to the bootstrap stack.
+- Implement the actual refactoring.
+- Implement rollback.
+- Implement the progress bar to display the refactoring progress.
+- Implement feature flags.
+- Handle cross-stack references.
+- Add the refactor step to the `deploy` command.
+- Write a blog post.
+
+### Are there any open issues that need to be addressed later?
+
+This improved deployment experience actually consists of two separate steps,
+behind the scenes: refactoring followed by deployment. And the whole workflow is
+controlled by the CLI. As a result, this is not an atomic operation: it is
+possible that the refactoring step succeeds, but before the CLI has a chance to
+deploy the changes, it gets interrupted (computer crash, network failures, etc.)
+In this case, the user will be left with a stack that is neither in the original
+nor in the desired state.
+
+In particular, the logical ID won't match the CDK construct path, stored in the
+resource's metadata. This has consequences for the CloudFormation console, which
+will show a Tree view that is not consistent with the Flat view.
+
+Possible solutions to consider, from more specific to more general:
+
+- CloudFormation to ignore changes in the `Metadata[aws:cdk:path]` resource
+  attribute in refactor operations.
+- CloudFormation to allow resource additions and deletions in refactor
+  operations.
+- Two-phase commit. The CLI could create the refactor and the changeset, and
+  then have a new command to execute both in a single atomic operation (let's
+  say, a `executeChangeSetAndRefactor()`).
+
+Since all these options depend on changes on the CloudFormation side, and this
+edge case is unlikely to happen, we are going to address it later.
+
+## Appendix
+
+### A. Equivalence between resources
+
+To detect which resources should be refactored, we need to indentify which
```

Comment:
So we don't check for resources that will be deleted and ones that will be created and find some near equivalency? Could we generate a warning, e.g. Lambda A is being deleted Lambda B is being created, one property difference is preventing refactoring.

Reply:
I am very much interested in that generalization: replacing the notion of equivalence with the notion of distance between two resources. Equivalent resources would then be a special case (distance = 0). The challenge is to come up with a workable distance function, and probably also a threshold. But given the interest (this is also related to Shikha's question), I can add this to the scope, but address in a third phase.